### DISCUSSION TITLE
ensure order of tests whenever using a dataProvider

### DISCUSSION THREAD

**Discussion by bitcoder** [cbeust/testng#3109] [Q&A] [answered]
Hi everyone,
I want to ensure that tests generated using a DataProvider, as shown ahead, are executed by order.
The execution order seems that is not guarantee (sometimes it executes in one way, sometimes in another way); in the docs I wasn't able to find an explicit statement  but I did find [this issue](https://github.com/testng-team/testng/issues/799) which may (?) be related.
Anway, I want to ensure the items returns by the DataProvider to be executed by order. How can I ensure that? Do I need to refactor this code somehow?
Thanks in adbaance.

```java
public class DataDrivenExamples {
 
    @DataProvider(name = "numbers")
    public static Object[][] evenNumbers() {
        // the last data/example if not correct on purpose, to fail the test
        return new Object[][]{{1, false}, {2, true}, {4, true}, {5, true}};
    }
  

    @Test(dataProvider = "numbers")
    public void givenNumberFromDataProvider_ifEvenCheckOK_thenCorrect(Integer number, boolean expected) {
        assertEquals(expected, number % 2 == 0);
    }

}
```

Note: tested using TestNG 7.9.0

**Comment by itkhanz** [comment]
@krmahadevan could you please have a look at this discussion and sample code-snippet and suggest if there's a way to preserve order of tests in this scenario?

**Reply by krmahadevan** [reply]
@itkhanz - The sample is not complete. Can you or @bitcoder please share a sample that can demonstrate the scenario wherein TestNG is executing things out of order?

@bitcoder - I dont see anything in the TestNG codebase that can lead to altering the order of execution of a test method powered by a data provider (Factories can be a totally different case and their entire behavior is different since there is test class instantiation involved). 

So can you share a sample that I can use to see TestNG executing the tests out of order. The only way in which this is possible is when you have enabled method level parallelism (or) if you are trying to do this with a `@Factory` instantiated test method classes.

Either case, please use `7.10.1` and share a sample that can be used to recreate the discrepancy you are noticing.

**Reply by krmahadevan** [reply]
Here's a sample that I could think of which demonstrates that TestNG executes a data driven test, with parameters provided by a data provider, in the exact same way they were producted

```java
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;

public class DataDrivenExamples {

    private final AtomicInteger counter = new AtomicInteger(1);

    @DataProvider
    public Object[][] getData() {
        return new Object[][]{
                {false}, {true}, {false}, {true}, {false}
        };
    }

    @Test(dataProvider = "getData")
    public void testOrder(boolean expected) {
        try {
            assertThat(counter.get() % 2 == 0)
                    .withFailMessage(counter.get() + " should have been " + text(expected))
                    .isEqualTo(expected);
        } finally {
            counter.incrementAndGet();
        }
    }

    private static String text(boolean flag) {
        return (flag) ? "even" : "odd";
    }
}
```

**Reply by itkhanz** [reply]
Thank you @krmahadevan for details response. We solved our issue by sorting the results.

Below you will find the details of ticket which was causing flakiness of the build due to unexpected ordering of DataDriven Tests:
https://github.com/Xray-App/xray-testng-extensions/issues/12

This is the test that was failing due to flakiness
https://github.com/Xray-App/xray-testng-extensions/blob/828fe56cb406e66b2aa2be1958fcd19ff17ba578/src/test/java/app/getxray/xray/testng/tests/XrayJsonReportTests.java#L197

**Reply by krmahadevan** [reply]
@itkhanz - Thanks for getting back. But I would still like to see a sample that can be used to reproduce the problem. So if you or @bitcoder can help share a test (or) alter the one that I shared earlier, I would still want to investigate as to what is going wrong.

**Reply by itkhanz** [reply]
This is the test case that was failing earlier because it expect the response in TestNG report to be in same order as the parameters in data driven scenario, and sometimes the order of iterations were different making the build flaky. We fixed it at our end by sorting the iterations and then making comparison.
https://github.com/Xray-App/xray-testng-extensions/blob/master/src/test/java/app/getxray/xray/testng/tests/XrayJsonReportTests.java#L195

and this is the data-driven scenario class and method that is being run
https://github.com/Xray-App/xray-testng-extensions/blob/828fe56cb406e66b2aa2be1958fcd19ff17ba578/src/test/java/app/getxray/xray/testng/tests/DataDrivenExamples.java#L26C1-L42C6

This is how the test class and method name are then being passsed to run the test:
https://github.com/Xray-App/xray-testng-extensions/blob/828fe56cb406e66b2aa2be1958fcd19ff17ba578/src/test/java/app/getxray/xray/testng/tests/XrayJsonReportTests.java#L509

**Reply by krmahadevan** [reply, marked as answer]
@itkhanz - The root cause of the issue is NOT how TestNG runs the tests but its related to how TestNG captures the results. 

I see that the code does use a `LinkedHashSet` to sort the `ITestResult` objects based on the [start time](https://github.com/Xray-App/xray-testng-extensions/blob/master/src/main/java/app/getxray/xray/testng/listeners/XrayJsonReporter.java#L500-L508) 

But this is being persisted into `results` which is declared as a `HashMap` instead of it being a `LinkedHashMap`. So the ordering is being lost. See [here](https://github.com/Xray-App/xray-testng-extensions/blob/master/src/main/java/app/getxray/xray/testng/listeners/XrayJsonReporter.java#L75)

I think [this code block](https://github.com/Xray-App/xray-testng-extensions/blob/master/src/main/java/app/getxray/xray/testng/listeners/XrayJsonReporter.java#L189-L201) could have been just replaced with something like this (which would ensure that the ordering is going to be maintained)

```java
Map<String, List<ITestResult>> results = testResults.stream()
        .collect(
                Collectors.groupingBy(
                        itr -> itr.getMethod().getQualifiedName(),
                        LinkedHashMap::new,
                        Collectors.toList()
                )
        );
```

This will ensure that the entries getting added into `results` maintain the insertion order. So when you iterate this map and then add them into the JSONArray, the order in which they were found should be preserved.

```
results.values().forEach(it -> addTestResults(tests, it));
```

In a nutshell, I think your tests were right and they were actually hinting at a problem in the listener code which should have been fixed 😬

**Reply by krmahadevan** [reply]
@bitcoder @itkhanz - Let me know if that provides a convincing explanation to what could be going wrong in this case.

**Reply by bitcoder** [reply]
thanks @krmahadevan  for your analysis and your time; some time ago I thought also the same (that it was due to the current listener implementation); I will have a look at it whenever I have a bit and I'll let you know
